librz/core/linux_heap_glibc.c

```diff
@@ -1485,15 +1553,239 @@ void GH(print_malloc_info)(RzCore *core, GHT m_state, GHT malloc_state) {
+		return 0;
+	}
+
+	while (fw != head->fd) {
+		rz_io_read_at(core->io, fw, (ut8 *)cnk, sizeof(GH(RzHeapChunk)));
+		rz_cons_printf(" -> ");
+		GH(print_heap_chunk_simple)
+		(core, fw);
+		fw = cnk->fd;
+		chunks_cnt += 1;
+	}
+	rz_cons_newline();
+	free(cnk);
+	free(head);
+
+	return chunks_cnt;
+}
+
+/**
+ * \brief Prints unsorted bin description for an arena (used for `dmhd` command)
+ * \param core RzCore pointer
+ * \param m_arena Offset of the arena in memory
+ * \param main_arena MallocState struct for the arena in which bin are
+ */
+static void GH(print_unsortedbin_description)(RzCore *core, GHT m_arena, MallocState *main_arena) {
+	RzConsPrintablePalette *pal = &rz_cons_singleton()->context->pal;
+	rz_cons_printf("Unsorted bin @ ");
+	PRINTF_BA("0x%" PFMT64x "\n", (ut64)m_arena);
+	int chunk_cnt = GH(print_bin_content)(core, main_arena, 0);
+	rz_cons_printf("Found %d chunks in unsorted bins \n", chunk_cnt);
+}
+
+/**
+ * \brief Prints small bins description for an arena (used for `dmhd` command)
+ * \param core RzCore pointer
+ * \param m_arena Offset of the arena in memory
+ * \param main_arena Pointer to MallocState struct for the arena in which bins are
+ */
+static void GH(print_smallbin_description)(RzCore *core, GHT m_arena, MallocState *main_arena) {
+	RzConsPrintablePalette *pal = &rz_cons_singleton()->context->pal;
+	rz_cons_printf("Small bins @ ");
+	PRINTF_BA("0x%" PFMT64x "\n", (ut64)m_arena);
+	int chunk_cnt = 0;
+	int non_empty_cnt = 0;
+	for (int bin_num = 1; bin_num < NSMALLBINS; bin_num++) {
+		int chunk_found = GH(print_bin_content)(core, main_arena, bin_num);
+		if (chunk_found > 0) {
+			non_empty_cnt += 1;
+		}
+		chunk_cnt += chunk_found;
+	}
+	rz_cons_printf("Found %d chunks in %d small bins \n", chunk_cnt, non_empty_cnt);
+}
+
+/**
+ * \brief Prints large bins description for an arena (used for `dmhd` command)
+ * \param core RzCore pointer
+ * \param m_arena Offset of the arena in memory
+ * \param main_arena Pointer to MallocState struct for the arena in which bins are
+ */
+static void GH(print_largebin_description)(RzCore *core, GHT m_arena, MallocState *main_arena) {
+	RzConsPrintablePalette *pal = &rz_cons_singleton()->context->pal;
+	rz_cons_printf("Large bins @ ");
+	PRINTF_BA("0x%" PFMT64x "\n", (ut64)m_arena);
+	int chunk_cnt = 0;
+	int non_empty_cnt = 0;
+	for (int bin_num = NSMALLBINS; bin_num < NBINS - 2; bin_num++) {
+		int chunk_found = GH(print_bin_content)(core, main_arena, bin_num);
+		if (chunk_found > 0) {
+			non_empty_cnt += 1;
+		}
+		chunk_cnt += chunk_found;
+	}
+	rz_cons_printf("Found %d chunks in %d large bins \n", chunk_cnt, non_empty_cnt);
+}
+
+/**
+ * \brief Prints tcache description for main arena (used for `dmhd` command)
+ * \param core RzCore pointer
+ * \param m_arena Offset of main arena in memory
+ * \param main_arena Pointer to MallocState struct for main arena
+ */
+static void GH(print_tcache_description)(RzCore *core, GHT m_arena, MallocState *main_arena) {
```

Comment:
There was already some code that printed `tcache` (see `dmht`). Would it make sense to use the same code instead of "duplicating"? Or is this very different from the other code?

Reply:
I think with a slight modification it is possible to reuse `dmht` code. I will do that.

Reply:
I didn't check, but if there's other code that can be shared for the other types of bins, please do so. It helps to avoid inconsistencies.

Reply:
I think `dmhf` code can also be integrated